gadopt/stokes_integrators.py

```diff
@@ -140,14 +155,64 @@ def __init__(self, z, T, approximation, bcs=None, mu=1,
                     self.strong_bcs.append(fd.DirichletBC(self.Z.sub(0).sub(1), value, id))
                 elif bc_type == 'uz':
                     self.strong_bcs.append(fd.DirichletBC(self.Z.sub(0).sub(2), value, id))
+                elif bc_type == 'free_surface':
+                    # Copy free surface information to a new dictionary
+                    free_surface_dict[id] = value
+                    self.free_surface = True
                 else:
                     weak_bc[bc_type] = value
             self.weak_bcs[id] = weak_bc
 
+        if self.free_surface:
+            if free_surface_dt is None:
+                raise TypeError("Please provide a timestep to advance the free surface, currently free_surface_dt=None.")
+
+            self.solution_old = fd.Function(self.solution)
```

Comment:
Since this comment @Patol75  added a solution old too. We discussed it a little bit yesterday but I'm not 100 % sure why you have added it in Thomas?